### PR TITLE
Add Intersperse for lists

### PR DESCRIPTION
**Pull request by markuspf**

### CODE REVIEW

**Comment by alex-konovalov** [comment]
Looks like a useful utility. For example, `Intersperse( "abc", ',' );` returns `"a,b,c"`.

File: lib/list.gd [hunk 1]
```diff
@@ -1661,6 +1661,29 @@ DeclareOperation( "SortParallel",
...
+#############################################################################
+##
+#F  Intersperse( <list>, <elt> )  . . . . . . . . intersperse list with elt
+##
+##  <#GAPDoc Label="Intersperse">
+##  <ManSection>
+##  <Oper Name="Intersperse" Arg='list, elt'/>
+##
+##  <Description>
+##  Returns a new mutable list that is interspersed that consists of
```
Comment (alex-konovalov) [inline lib/list.gd]: Is there a word missing here? The text doesn't flow.

**Comment by fingolfin** [comment]
What are suggested real-world applications for this? Alex' example doesn't convince me at all. More realistiv variants of it that I used in practice were solved just fine by using `JoinStringsWithSeparator`.

My point here is: I'd only want to add something like this to GAP if it has an actual usecase. Otherwise, it just ends up polluting the namespace. At the very least, a pointer to any other language which has such a function in their standard library would be of interest...

**Comment by fingolfin** [comment]
Put another way: If you were adding a cool new semi group algorithm to GAP, which needs this function, that would convince me. Adding the function in isolation, OTOH, seems arbitrary.

BTW, if we add this, it would be nice if there was a test case (but of course it doesn't make sense to work on that before there's a decision).

**Comment by ChrisJefferson** [comment]
Personally, I wouldn't have known to look for "JoinStringsWithSeperator". The two functions are fairly similar. It might be worth looking at any use-cases and considering extending JoinStringsWithSeperator, and/or add a new index reference for 'Intersperse' if anyone is looking under that name.

**Comment by fingolfin** [comment]
Thing is, many very popular languages have a "join" function for strings resp. lists (e.g. JavaScript, Perl, Python, Ruby, C#, ...), so it's fairly well-known idiom; the name in GAP is a bit awkward, but quite easy to discover using TAB completion.

I am not aware of any languages having an "intersperse" function, certainly none of the "big" ones.

**Comment by alex-konovalov** [comment]
I think we're diverting into applying this to strings only - note that `Intersperse` is applicable to arbitrary lists. BTW, intersperse exists in haskell (see http://hackage.haskell.org/package/base-4.8.1.0/docs/Data-List.html#v:intersperse) but was not accepted in Python (see https://bugs.python.org/issue21414)

**Comment by rbehrends** [comment]
> I am not aware of any languages having an "intersperse" function, certainly none of the "big" ones.

[Haskell](https://hackage.haskell.org/package/base-4.8.1.0/docs/Data-List.html) and [OCaml](https://ocaml.janestreet.com/ocaml-core/111.28.00/doc/core/#Core_list) (the latter via Core.Std) do. It's more of a functional idiom, so it's less common in imperative languages.

**Comment by fingolfin** [comment]
@alex-konovalov Sure, join is limited to strings, where I have used it frequentl. OTOH, I have not been able to think of a past programming task I worked on where I ever wanted something like intersperse. Of course that doesn't mean it's not useful for somebody. But then surely somebody can provide an actual, non-contrived usecase where it would be beneficial?

@rbehrends I see. Well, Haskell and OCaml certainly are not obscure languages at all. So, what functional use cases involve intersperse?

**Comment by rbehrends** [comment]
@fingolfin It is more that functional languages are more likely to provide reusable list transformations and that imperative languages are more willing to do them in an ad-hoc fashion (in my experience, at least). An example of an application that uses `intersperse` frequently is Pandoc, which applies it to lists of tokens or ASTs.

**Comment by markuspf** [comment]
For my current usecase JoinStringsWithSeparator is indeed enough, but I didn't find the bugger (or I did at some point, and did not remember).
Indeed I know this idiom from functional programming, but since GAP is an imperative language (and I believe we should even aim for Python similarity as much as we can) it'd be more sensible to go with whatever is common in imperative world.

**Comment by rbehrends** [comment]
@markuspf I didn't mean to imply that this is functional-only. Functional languages simply are more likely to have such abstractions, because it beats having to write a tail-recursive function each time (and verifying that it's tail-recursive and won't blow up the stack). It's much easier to do this ad-hoc with loops, but it still violates DRY.

The other reason that you see it rarely with interpreted languages is that (unfortunately, IMO) interpreted languages eschew defining new types (because wrapping data in types incurs measurable CPU overhead) and tend to overly rely on the basic types (typically, ints, floats, strings, lists, hash tables). While that is really not good for abstraction, it also means that you end up having less need for generic functions.

In general, while I wouldn't go to the mat for `intersperse` specifically, I think that GAP could benefit from a better set of basic library routines; adopting libraries from Haskell and OCaml (so that there's a set of more powerful primitives available when needed) might be a good idea. There's a much stronger argument for a full library of list routines than for any given list function.

**Comment by ChrisJefferson** [comment]
@rbehrends : If we are planning a new list library, I've always wished that more of GAP's primitives worked around enumerators, that is something like:

`for i in List([1..10000], AlternatingGroup) do ... od`, shouldn't actually ever create a list of 10,000 groups, but create the groups as required during the iteration.

GAP already has the infrastructure to do this, but it isn't very well integrated or documented.

**Comment by frankluebeck** [comment]
I also cannot see any use case for `Intersperse`. Even if there was one package using this somewhere it would not convince me that it should go into the library (or that there is a specific function for this at all). 

More generally, I would add such trivial functions only if a regular use by several users is to be expected. Otherwise, such functions eat up namespace and make it more difficult to find the more useful functions.

ChrisJefferson said:
> `for i in List([1..10000], AlternatingGroup) do ... od`, shouldn't actually ever create a list of 10,000 groups, but create the groups as required during the iteration.

I don't agree with this. When GAP parses the part `List([1..10000], AlternatingGroup)` it does exactly what you ask for, namely it creates 10000 groups. One could easily implement an enumerator such that `for i in Enumerator([1..10000], AlternatingGroup) do ... od` works without creating a list of 10000 groups. But even that looks unnecessarily complicated to me because one can just do
`for n in [1..10000] do i := AlternatingGroup(n); ... od`.

**Comment by stevelinton** [comment]
@frankluebeck I agree about changing the specific behaviour of List, but I think Chris's general point -- that we should promote use of iterators (which is think is what Chris means) in our designs where we can. In general, if we are writing a function that logically returns a finite sequence of things, we should start by thinking of a version that produces an iterator and allow anyone who wants a random access list to make one from the iterator. 

If we were starting over, then simple names like "Filtered" or "Cartesian" should probably produce the iterators rather than flat lists, but that boat has sailed. 

What we could do now is systematically add functions with names like ListIterator (or maybe just Iterator) FilteredIterator, etc. We could also extend the input specification of these functions to accept Iterators as input and perhaps provide slightly nicer printing for Iterators.

**Comment by fingolfin** [comment]
@stevelinton Depends on wht you mean with "nice way to take a few comments"... in this particular case, I would just open a new issue and copy & past relevant bits, possibly using that one can mark text passages as quotes by prefixing them with >, like this:
> a quote
But things like "making more use of iterators" maybe should discussed on the mailing list instead, and issues reserved for more concrete things? 

Anyway, this is quite off-topic now. I still think that adding `intersperse` to GAP like this would be weird. And I say that despite being a big fan of functional patterns -- but before we add such a rather special and (to me) exotic one, I'd prefer if we first added things like fold, etc., and in a consistent way. But that, too, is a separate discussion, which I'd be happy to have on the mailing list or during the next GAP days!

But I think we should close this particular issue.

**Comment by markuspf** [comment]
I will close this issue here, and encourage general discussion of libraries and functionality on the mailing list, as that is the correct place for such a thing (in my opinion).

**Pull request closed without merging**